Update API RFD. (#6764)
gravitational/teleport: rfd/0010-api.md

## What

- Unify Golang Client experience.
- Automatically generate API documentation using [pkg.go.dev](https://pkg.go.dev/).

## Why

There are several problems with the Teleport Go client libraries:

* The [Go client example](https://github.com/gravitational/teleport/tree/branch/5.0/examples/go-client) uses the teleport Go module as a dependency. This means that the example pulls in all of the teleport module's packages and indirect dependencies, leading to extreme import bloat.

> * There is no separate client library with guarantees of compatibility with specific version of Teleport. See examples [here](https://github.com/kubernetes/client-go#compatibility-matrix).

Reviewer: ```suggestion
* There is no separate client library with guarantees of compatibility with a specific version of Teleport. See examples [here](https://github.com/kubernetes/client-go#compatibility-matrix).
```

    // TLS client is the only client
    return auth.NewTLSClient(clientConfig)
}
```

## Details

A new `api` package will be made these new features and any existing Teleport Auth client logic, most of which is is currently in `teleport/lib/auth/clt.go`. The new package will only support Teleport's gRPC client which will gradually replace the deprecated HTTP client. Other logic such as the web client and certificate reading logic will also be moved to the `api` package. 

> The `api` package will be made a Go module, with no dependency on Teleport libraries. This will make it easy to import the API client with a small import footprint. Users will be able to run `go get github.com/gravitational/teleport/api` to get the latest version of the client and get started.

Author: Yes, that's a good point. I think we've pretty much settled on keeping the api a go sub module since the main repo has direct dependencies on it. It would be a pain to develop in separate repositories since most changes require both modules to be changed. The more likely long term solution in my mind is to turn more teleport packages that aren't purely internal (e.g. `lib/utils`, `lib/defaults`) into submodules. I've seen a few large scale projects like https://github.com/hashicorp/vault use this approach. This will be difficult due to the convoluted package dependency graph within the teleport module though.

##### Deprecate the `lib/auth.Client`

As stated before, all gRPC client related code will be moved out of `teleport/lib/auth/clt.go` and into `api/client`. Backwards compatibility will be maintained by embedding the gRPC `api/client.Client` in the internal `teleport/lib/auth.Client`.

Additionally, the remaining HTTP endpoints in the internal client will be converted and added to the gRPC client. To maintain backwards compatibility with old servers, the internal client will be updated to first try the new gRPC endpoint before falling back to the existing HTTP endpoint. These fallback methods will be moved to `teleport/lib/auth/httpfallback.go` and deprecated after one major release cycle.

There is [issue](https://github.com/gravitational/teleport/issues/6394) used to track remaining endpoints that need to be converted to gRPC. 

Once all necessary endpoints have been added in gRPC and all HTTP fallback logic has been deprecated and removed, the gRPC client can completely replace the existing `lib/auth.Client`. At this point the underlying gRPC server will be the same, so the transition should be relatviley straightforward.

### api module

> The api package will be made a sub module of teleport. It will exist in the `teleport/api`, and will be imported by the main `teleport` module through a go.mod [replace directive](https://golang.org/ref/mod#go-mod-file-replace) using a relative path.

Reviewer: ```suggestion
The api package will be made a sub module of teleport. It will exist in the `teleport/api` directory, and will be imported by the main `teleport` module through a go.mod [replace directive](https://golang.org/ref/mod#go-mod-file-replace) using a relative path.
```

Users will be able to get the latest api version by using `go get github.com/gravitational/teleport/api`, or a specific version with `go get github.com/gravitational/teleport/api@vX.Y.Z`.

#### versioning

Go module versioning must follow semver. Since [Teleport versioning](https://github.com/gravitational/teleport/blob/master/rfd/0012-teleport-versioning.md) is a subset of semver, the API can easily follow both version requirements by matching the Teleport version. This means that when `v7.0.1` of Teleport is released, `v7.0.1` of the API will also be released.

> Go module versions can automatically be pushed with git tags. Since the api module is a sub module, its version tags will look like `git tag api/vX.Y.Z`. The version can be tagged as part of the release process just like the teleport repository.

Reviewer: Is this strictly necessary if the API version is always going to be equal to the Teleport version?

Perhaps I'm misunderstanding, but I feel like we'd be better off updating our Teleport `Makefile` logic to just update the version declared in the API's `go.mod` file whenever we do a Teleport release and bundle it under the same tag as Teleport.

Needing to have two separate tags would increase the cognitive load on the release engineer, as the entire release process is still mostly very manual.

Author: Unfortunately, yes - https://github.com/golang/go/wiki/Modules#is-it-possible-to-add-a-module-to-a-multi-module-repository

As an example, if you look at at [hashicorp vault](https://github.com/hashicorp/vault), you'll see that `vault/api` and `vault/sdk` are tagged this way, which directly leads to go catching their versions - https://pkg.go.dev/github.com/hashicorp/vault/api?tab=versions

Actually, this looks a bit more complicated though. The main `teleport` module actually isn't following go module rules right now, so it might qualify as a package for the first scenario in the first link above. I'll test this out.

Additionally, we could choose to release the client out of step with Teleport releases. One option would be to release the api as `v1.[major-teleport-version].[api patch/minor version]`. This would allow us to avoid the need for the `vX` suffix, and make it easy to release the api separately from teleport (when needed) while still maintaining the teleport versioning guidelines in a sense (api v1.7.X should work with teleport v7.Y.Z). However solutions such as this move the cognitive load off of release management and onto customers/developing/marketing. But if we want to explore more options like this I can list out some ideas with pros/cons.

Author: I tried testing the tags for versioning - https://pkg.go.dev/github.com/Joerger/module-versioning-test.

It looks like submodules have to be tagged separately in our case. Otherwise the [submodule](https://pkg.go.dev/github.com/Joerger/module-versioning-test/submodule/v2) would have a v2.0.0 version available.

Reviewer: I see what you mean.

The interesting thing with the Hashicorp example is that they don't release a new SDK/API tag whenever they release a new version of Vault - so while they have lots of `vx.x.x` tags for Vault itself (as Teleport does) they only have a handful of API/SDK tags like `sdk/v1.1.0` and `api/v1.1.0` which makes the tag list much cleaner.

My personal opinion is that we should handle tagging in a similar way to Vault. Although there will be some burden on customers to use the correct API version for their Teleport installation, it should actually make it easier for them overall as they might be able to upgrade through several Teleport versions without needing to change the version of the API they use. We're unlikely to make breaking public API changes unless we absolutely need to, right?

Also, this would definitely be the correct time for us to make this decision as we don't have any public releases of our API at the moment. We could implement everything and release a tag starting off at `v1`, or use the major/minor of Teleport as you suggested.

I think it's worth exploring some different options other than just bumping the API version every time we release Teleport.

Author: @webvictim I flushed out the two release options, I think Release Option 1 is actually going to put less burden on release managers, but let me know what you think.

    // TLS client is the only client
    return auth.NewTLSClient(clientConfig)
}
```

## Details

A new `api` Go module will be made to expose existing and new API features. It will be a submodule of `teleport`, so it will be impossible to have `teleport` as a dependency. This will make it easy to import the API with a small import footprint. Users will be able to run `go get github.com/gravitational/teleport/api` to get the latest version of the API and get started.

Teleport's internal gRPC Auth client will be moved into the `api/client` package and embedded in the old client for backwards compatibility. Over time, the internal client will be deprecated and replaced with the new public client.

Other logic such as the web client and tls/ssh certificate logic will be moved to the `api` package as well.

> The new client will gain several new quality of life improvements, including credential loaders, automatic server version checking, better documentation, and more language implementations.

Author: IIUC the `TeleportClient` used by `tsh` has this, but the Auth client in `lib/auth/clt.go` does not have this.

There is [issue](https://github.com/gravitational/teleport/issues/6394) used to track remaining endpoints that need to be converted to gRPC. 

Once all necessary endpoints have been added in gRPC and all HTTP fallback logic has been deprecated and removed, the gRPC client can completely replace the existing `lib/auth.Client`. At this point the underlying gRPC server will be the same, so the transition should be relatviley straightforward.

### api module

The api package will be made a sub module of teleport. It will exist in the `teleport/api`, and will be imported by the main `teleport` module through a go.mod [replace directive](https://golang.org/ref/mod#go-mod-file-replace) using a relative path.

Users will be able to get the latest api version by using `go get github.com/gravitational/teleport/api`, or a specific version with `go get github.com/gravitational/teleport/api@vX.Y.Z`.

#### versioning

> Go module versioning must follow semver and can automatically be pushed with git tags. Since the api module is a sub module, its version tags will look like `git tag api/vX.Y.Z`.

Reviewer 2: why do we need a separate tag for `api`?

Author: Unfortunately It's required - https://github.com/golang/go/wiki/Modules#is-it-possible-to-add-a-module-to-a-multi-module-repository

Gus asked the same questsion - https://github.com/gravitational/teleport/pull/6764#discussion_r631248431

##### Compatibility

The gRPC API and other code imported by `github.com/gravitational/teleport` must follow the [Teleport version compatibility](https://github.com/gravitational/teleport/blob/master/rfd/0012-teleport-versioning.md#compatibility) rules. However, the api module itself does not need to follow the same versioning rules. For example, if the API module has a new feature that is not used by the main teleport module, such as the Credential Loaders, it does not need to be compatible with previous versions of Teleport.

This gives us some flexibility to choose an api versioning strategy without worries of unrelated versioning promises.

##### Release option 1 - Release the API with each Teleport release

The API can be tagged for release as part of the existing Teleport release process. Each Teleport release will get both the `vX.Y.Z` and `api/vX.Y.Z` tags.

This can be done within the current [release checklist](https://gravitational.slab.com/posts/teleport-release-checklist-cf0s6vmt) with just a few additional steps:
 - Add git tag - `api/vX.Y.Z`
 - If it's a major version bump...
   - Update `api/go.mod` with the correct `vX` major version suffix.

>    - Update `teleport/go.mod` to import the new version suffix and update any `teleport/api` imports with the updated suffix.

Reviewer: We should ideally automate these steps as much as possible to make sure that they aren't missed from a release.

   - Since the client will have automatic server version checks, users should be able to quickly catch incorrect version usage regardless.
 - Release management - The API will need to be released on its own schedule with its own versions and branches. This will burden release managers with basically a new product to worry about. 
   - In the long run, this may be less of a burden than it might seem at first. With `Release option 1`, we may have 20 or so new versions of `teleport/api` released, while with `Release option 2`, most likely there will only be a couple of new versions.

This option provides the most flexibility in API releases. It also gives us the option to start with `v0`, with no stability/compatibility promises, in order to try out a strategy before committing fully to it. If treating the API as a separately versioned product is too burdensome, we can immediately transition to another release strategy.

##### Release option choice

> In my opinion, `Release option 1` is the better choice. It will be much easier to get started with its versioning since it can follow all the same release processes in place. `Release option 2` brings many more complexities into the discussion, such as new compatibility rules for interfacing with the API and new branches and tags for the api. 
> 
> However, it may be that releasing the API and other futures modules separately (Option 2) will give less headaches at scale, but this is hard for me to measure. For that reason, it would make more sense for the release managers to choose a strategy. So far the consensus is to go with Release Option 1.

Reviewer 2: ```suggestion
The consensus is `Release option 1`.

It will be much easier to get started with its versioning since it can follow all the same release processes in place. `Release option 2` brings many more complexities into the discussion, such as new compatibility rules for interfacing with the API and new branches and tags for the api. 
```

Additionally, the remaining HTTP endpoints in the internal client will be converted and added to the gRPC client. To maintain backwards compatibility with old servers, the internal client will be updated to first try the new gRPC endpoint before falling back to the existing HTTP endpoint. These fallback methods will be moved to `teleport/lib/auth/httpfallback.go` and deprecated after one major release cycle.

There is [issue](https://github.com/gravitational/teleport/issues/6394) used to track remaining endpoints that need to be converted to gRPC. 

Once all necessary endpoints have been added in gRPC and all HTTP fallback logic has been deprecated and removed, the gRPC client can completely replace the existing `lib/auth.Client`. At this point the underlying gRPC server will be the same, so the transition should be relatviley straightforward.

### api module

The api package will be made a sub module of teleport. It will exist in the `teleport/api`, and will be imported by the main `teleport` module through a go.mod [replace directive](https://golang.org/ref/mod#go-mod-file-replace) using a relative path.

> Users will be able to get the latest api version by using `go get github.com/gravitational/teleport/api`, or a specific version with `go get github.com/gravitational/teleport/api@vX.Y.Z`.

Author: Yes, mainly because part of our versioning promise is to maintain compatibility between "binaries", which includes the API. It's a lot easier to reason and maintain if it follows the same exact versioning. The main issue with this approach is that we may have extraneous releases with no actual API changes, in which case we could consider skipping some minor/patch releases for the API. I think major/minor versions should always line up though.

> Users will be able to get the latest api version by using `go get github.com/gravitational/teleport/api`, or a specific version with `go get github.com/gravitational/teleport/api@vX.Y.Z`.

Reviewer 2: we guarantee compatibility for binaries and gRPC API but we won't guarantee compatibility of the Go API under `api/`.
so while the gRPC API should remain compatible between v6.1 and v6.2, we can make it acceptable to break external Go code that depends on `teleport/api` when upgrading from v6.1 to v6.2 of the module.

Since Teleport has a major version above 2, the api must follow the [v2+ Go Module release guide](https://github.com/golang/go/wiki/Modules#releasing-modules-v2-or-higher) by appending a major version suffix to the import path. The import path for each major version `X` will look like `github.com/gravitational/teleport/api/vX`.

This import path needs to be updated within the entire teleport repository. This can be done before each major release. This import path change is likely to cause minor merge conflicts in backport PRs that are a full major version back.

> Additionally, if the git tag and import suffix aren't set up, the api will be versioned as v0.0.0-[random-string]. Getting the latest version with `go get github.com/gravitational/teleport/api` will get the most recent version in terms of time rather than semver/teleport version, confusing and not user friendly. For this reason it's important to start tagging the version and add the `vX` suffix as soon as possible.

Reviewer 3: So, for example, the command `go get github.com/gravitational/teleport/api/v7@v9.1.0` would mean "get me the Teleport 9.1 implementation of API version 7"? 

(Just an example, not that such a thing has to exist given our backwards compatibility rules)

Author: Actually I believe Go will return an error on that call. The major version suffix and the major version must line up, so the version suffix would have to be updated every major release.

The API's documentation will be split into 3 parts:
 - Automatically generated [pkg.go.dev webpage](https://pkg.go.dev/github.com/gravitational/teleport/api/client).
 - [API Reference](https://goteleport.com/docs/reference/api/) on the Teleport docs site.
 - In depth code [examples](https://github.com/gravitational/teleport/tree/branch/v6.2/examples/go-client) 

#### pkg.go.dev (godoc)

The majority of the API's documentation will be hosted on [pkg.go.dev](https://pkg.go.dev/github.com/gravitational/teleport/api/client). This documentation is automatically generated from the package code upon each package release. Having code-centric documentation will make it easier to maintain and update. It will also improve the experience of working with the code both internally and externally.

In order to improve this automatically generated documentation, code comments throughout the api package will be made thorough and user focused. There are several useful [tricks](https://pkg.go.dev/github.com/fluhus/godoc-tricks) to improve the formatting of the documentation, and even add runnable [code examples](https://pkg.go.dev/github.com/fluhus/godoc-tricks#example-Examples).

> In the beginning, only essential functionality for the client package will have exceptional documentation. For example, [client.Credentials](https://pkg.go.dev/github.com/gravitational/teleport/api/client#Credentials) and [client.New](https://pkg.go.dev/github.com/gravitational/teleport/api/client#New) will have an extensive explanation of it's usage with code examples. 

Reviewer: ```suggestion
In the beginning, only essential functionality for the client package will have exceptional documentation. For example, [client.Credentials](https://pkg.go.dev/github.com/gravitational/teleport/api/client#Credentials) and [client.New](https://pkg.go.dev/github.com/gravitational/teleport/api/client#New) will have an extensive explanation of its usage with code examples. 
```

*spoiler*: The workflows example in `teleport-plugins/access/example` will be the first example to be added.

### Workflows

#### Deprecate teleport-plugins/access package

The `teleport-plugins/access` package was originally created in order to utilize the Teleport Auth API without being dependent on internal Teleport code. Now that Teleport has a external API, this is no longer a concern.

The access package is now basically a duplication of the api/client, with a few nonessential quality of life features over the `teleport/api/client` package:
  - Automatically append plugin name and other determinable parameters to relevant client methods. e.g. `GetPluginData(reqID)` calls `client.GetPluginData({types.KindAccessRequest, reqID, plugin.name})`.
  - Automatically filter and type cast access request events.
  - Rename, alias, or convert structs to be more user friendly for those who are only using access requests. e.g `types.AccessRequestV3` becomes `access.Request` which has a flat structure as opposed to `teleport.Resource` with metadata and spec structs embedded.

> These benefits are minimal at best, and actually detrimental at worst. Specifically, users who plan to use the access client for access request management the API client for other functionality will need to learn two APIs when they only need to learn one. Additionally, the listed benefits can be largely handled with improved documentation and examples.

Reviewer: ```suggestion
These benefits are minimal at best, and actually detrimental at worst. Specifically, users who plan to use the access client for access request management along with the API client for other functionality will need to learn two APIs when they only need to learn one. Additionally, the listed benefits can be largely handled with improved documentation and examples.
```

Another downside of the separate `access` package is that there are two separate clients to maintain, version/release. This adds unnecessary strain and confusion for both maintainers and users, especially when it comes to compatibility concerns.

The external plugins held in `teleport-plugins/access` will be updated to use the new client. Once this is complete, the access client can be removed, meaning its final version will be it's previous release. e.g. If we release the updated plugins in `v6.3`, The final version of the access client will be `v6.2`. `v6.2` would still be guaranteed to be compatible with teleport until `v8.0`, according to the Teleport [compatibility guidelines](https://github.com/gravitational/teleport/blob/master/rfd/0012-teleport-versioning.md#compatibility).

#### Add workflow examples and documentation

> In order to compensate for the removal of the `access` client and its qol features, the API client's documentation around workflows will be improved. User focused comments and examples will be added to all of the essential workflows functions, including the stream watcher.

Reviewer: ```suggestion
In order to compensate for the removal of the `access` client and its quality of life features, the API client's documentation around workflows will be improved. User focused comments and examples will be added to all of the essential workflows functions, including the stream watcher.
```

Access hooks are too low level and require in depth knowledge of teleport and the API client.

High level access hooks will be added to a new `api/client/access` in the form of a simple handler similar to an http server with a router.

```go
func myTeamRequest(ctx context.Context, req types.AccessRequest) (types.AccessRequestUpdate, error) {
  return DenyParams("No reason, just deny all."), nil
}

func main() {
router := access.Router()
  ctx := context.Background()

  router := access.Router{}
  // handle access requests from my team only
  router.HandleUpdateFunc(myTeamRequest, MatchUserTrait("team", "myteam"))
  srv := access.Server{
    Client: clt,
    Handler: router,
  }

  // Watch for incoming requests and automatically handle
  // any that match a HandleFunc.
  return srv.Run(ctx)
}
```

##### Route Matching

The router will use custom `Match` functions to match a request to a route.

```go
type Route struct {
	Handler  HandlerFunc
	Matchers []Match
}

// Custom function that will attempt to match a request
type Match func(req types.AccessRequest) bool

func MatchUserTrait(k, v string) Match { ... }
```

##### Matching Algorithm

> Unlike an http server router, the access hooks server cannot route via a radix tree. An access request may match multiple routes based on it's values. For example, a request may have `user.trait.team=myteam` and `user.label.group=security`. If both of these values have their own handler, then the request will match both handlers, despite the handlers having no detectable overlap.

Reviewer: ```suggestion
Unlike an http server router, the access hooks server cannot route via a radix tree. An access request may match multiple routes based on its values. For example, a request may have `user.trait.team=myteam` and `user.label.group=security`. If both of these values have their own handler, then the request will match both handlers, despite the handlers having no detectable overlap.
```

Since a simple tree-like solution is not an option, the router will instead loop over the routes available looking for a match with the request's associated matchers. The first one to match will be used to handle the request, so it is important for users to consider handler priority. 

##### HandleFunc

> Handler functions should be high level and easy to use for basic use cases. The most prevalent use case if a basic access request state update, so the base `HandlerFunc` will be designed to simply make a `client.SetAccessRequestState` request.

Reviewer: ```suggestion
Handler functions should be high level and easy to use for basic use cases. The most prevalent use case is a basic access request state update, so the base `HandlerFunc` will be designed to simply make a `client.SetAccessRequestState` request.
```

### Improve API documentation user experience

> Code comments throughout the api package will be made thourough and user focused. There will be examples and in depth explanations of user facing functions as needed. As a result, the automatically generated [pkg.go.dev](https://pkg.go.dev/github.com/gravitational/teleport/api/client) will be the main source of documentation.

Reviewer: ```suggestion
Code comments throughout the api package will be made thorough and user focused. There will be examples and in depth explanations of user facing functions as needed. As a result, the automatically generated [pkg.go.dev](https://pkg.go.dev/github.com/gravitational/teleport/api/client) will be the main source of documentation.
```
